### PR TITLE
Improve Ballerina MongoDB Connector

### PR DESCRIPTION
**Pull request by niveathika**
## Purpose
Usability issues such as passing collection names again and again in function calls and no support for interacting with multiple databases. 
Refer: https://groups.google.com/forum/?hl=en#!topic/ballerina-dev/z_amsHLV4Bk

## Goals
Improve usability by redesigning the API
Fix https://github.com/ballerina-platform/module-mongodb/issues/60

## Approach
Allowing users to access Client, Database and Collection as different objects

## User stories
https://docs.google.com/document/d/1mtwrhS_NnkI-spjQZyXefXTShZaaw5NqDvatWhsToQI/edit

## Release note
Improve Ballerina-MongoDB connector for better usability. The latest design focuses on allowing users to access Client, Database and Collection as different objects with improved function calls and param validation.

## Documentation
Attached in the PR

## Training
N/A

## Certification
N/A

## Marketing
N/A

## Automation tests
 - Unit tests - N/A
 - Integration tests - Added in the PR

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

## Samples
```ballerina
import ballerina/log;
import ballerina/mongodb;

public function main() {

    mongodb:ClientConfig mongoConfig = {
        host: "localhost",
        port: 27017,
        username: "admin",
        password: "admin",
        options: {sslEnabled: false, serverSelectionTimeout: 5000}
    };

    mongodb:Client mongoClient = checkpanic new (mongoConfig);
    mongodb:Database mongoDatabase = checkpanic mongoClient->getDatabase("Ballerina");
    mongodb:Collection mongoCollection = checkpanic mongoDatabase->getCollection("projects");

    map<json> doc1 = { "name": "ballerina", "type": "src" };
    map<json> doc2 = { "name": "connectors", "type": "artifacts" };
    map<json> doc3 = { "name": "docerina", "type": "src" };
    map<json> doc4 = { "name": "test", "type": "artifacts" };

    log:printInfo("------------------ Inserting Data -------------------");
    checkpanic mongoCollection->insert(doc1);
    checkpanic mongoCollection->insert(doc2);
    checkpanic mongoCollection->insert(doc3);
    checkpanic mongoCollection->insert(doc4);
  
    log:printInfo("------------------ Counting Data -------------------");
    int count = checkpanic mongoCollection->countDocuments(());
    log:printInfo("Count of the documents '" + count.toString() + "'.");


    log:printInfo("------------------ Querying Data -------------------");
    map<json>[] jsonRet = checkpanic mongoCollection->find(());
    log:printInfo("Returned documents '" + jsonRet.toString() + "'.");

    map<json> queryString = {name: "connectors" };
    jsonRet = checkpanic mongoCollection->find(queryString);
    log:printInfo("Returned Filtered documents '" + jsonRet.toString() + "'.");


    log:printInfo("------------------ Updating Data -------------------");
    map<json> replaceFilter = { "type": "artifacts" };
    map<json> replaceDoc = { "name": "main", "type": "artifacts" };

    int response = checkpanic mongoCollection->update(replaceDoc, replaceFilter, true);
    if (response > 0 ) {
        log:printInfo("Modified count: '" + response.toString() + "'.") ;
    } else {
        log:printInfo("Error in replacing data");
    }

   log:printInfo("------------------ Deleting Data -------------------");
   map<json> deleteFilter = { "name": "ballerina" };
   var deleteRet = checkpanic mongoCollection->delete(deleteFilter, true);
    if (response > 0 ) {
        log:printInfo("Delete count: '" + response.toString() + "'.") ;
    } else {
        log:printInfo("Error in deleting data");
    }

     mongoClient->close();
}
```

## Related PRs
N/A

## Migrations (if applicable)
Needed, refer README

## Test environment
Ballerina 1.2.0, MongoDB 4.2.0
 
## Learning
N/A

### CODE REVIEW

File: mongo-utils/src/main/java/org/wso2/mongo/MongoDBDataSourceUtil.java [hunk 1]
```diff
@@ -0,0 +1,287 @@
+// Copyright (c) 2019 WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (kalaiyarasiganeshalingam) [inline mongo-utils/src/main/java/org/wso2/mongo/MongoDBDataSourceUtil.java:1]:
```suggestion
// Copyright (c) 2020 WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```

File: mongodb/src/mongodb/client.bal [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    }
+
+    # List database names in the MongoDB server.
+    #
+    # + return - A array of database names on success else returns an error
+    public remote function getDatabasesNames() returns string[]|DatabaseError {
+        return getDatabasesNames(self.datasource);
+    }
+
+    #
```
Comment (kalaiyarasiganeshalingam) [inline mongodb/src/mongodb/client.bal:35]: Can you add the description?

File: mongodb/src/mongodb/collection.bal [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+    # List indices associated with the collection
+    #
+    # + return - json object with indices on success, else returns an error
+    public remote function listIndices() returns map<json>[]|DatabaseError {
+        return listIndices(self.collection);
+    }
+
+
+    # Insert one document
```
Comment (praneesha) [inline mongodb/src/mongodb/collection.bal:46]:
```suggestion
    # Inserts one document.
```

File: mongodb/src/mongodb/collection.bal [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+    #
+    # + document - Document to be inserted
+    # + return - () on success, else panics
```
Comment (praneesha) [inline mongodb/src/mongodb/collection.bal:49]:
```suggestion
    # + return - () on success or else panics
```

File: mongodb/src/mongodb/collection.bal [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+    public remote function insert(map<json> document) returns DatabaseError? {
+        string documentStr = document.toJsonString();
+        return insert(self.collection, java:fromString(documentStr));
+    }
+
+    # queries collection for documents and sorts and limit returned results.
```
Comment (praneesha) [inline mongodb/src/mongodb/collection.bal:55]:
```suggestion
    # The queries collection for documents, which sorts and limits the returned results.
```

File: mongodb/src/mongodb/collection.bal [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+    #
+    # + filter - Filter for query
```
Comment (praneesha) [inline mongodb/src/mongodb/collection.bal:57]:
```suggestion
    # + filter - Filter for the query
```

File: mongodb/src/mongodb/collection.bal [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+    # + sort - Sort options for query
```
Comment (praneesha) [inline mongodb/src/mongodb/collection.bal:58]:
```suggestion
    # + sort - Sort options for the query
```

File: mongodb/src/mongodb/collection.bal [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+    # + limit - Limit options for the query results. No limit is applied for -1
+    # + return - json array of the documents in the collection, else an error
```
Comment (praneesha) [inline mongodb/src/mongodb/collection.bal:60]:
```suggestion
    # + return - JSON array of the documents in the collection or else returns an error
```

File: mongodb/src/mongodb/collection.bal [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+        }
+        string filterStr = filter.toJsonString();
+        if (sort is ()) {
+            return find(self.collection, java:fromString(filterStr), (), limit);
+        }
+        string sortString = sort.toJsonString();
+        return find(self.collection, java:fromString(filterStr), java:fromString(sortString), limit);
+    }
+
+    # Update document based on condition
```
Comment (praneesha) [inline mongodb/src/mongodb/collection.bal:77]:
```suggestion
    # Updates a document based on a condition.
```

File: mongodb/src/mongodb/collection.bal [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+    #
+    # + set - Document for update condition
```
Comment (praneesha) [inline mongodb/src/mongodb/collection.bal:79]:
```suggestion
    # + set - Document for the update condition
```

File: mongodb/src/mongodb/collection.bal [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+    # + filter - Filter for query
```
Comment (praneesha) [inline mongodb/src/mongodb/collection.bal:80]:
```suggestion
    # + filter - Filter for the query
```

File: mongodb/src/mongodb/collection.bal [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+    # + isMultiple - Whether to update multiple documents
+    # + upsert - Whether to insert if update cannot be achieved
+    # + return - json array of the documents in the collection, else an error
```
Comment (praneesha) [inline mongodb/src/mongodb/collection.bal:83]:
```suggestion
    # + return - JSON array of the documents in the collection or else returns an error
```

File: mongodb/src/mongodb/collection.bal [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+    returns int|DatabaseError {
+        string updateDoc = set.toJsonString();
+        if (filter is ()) {
+            return update(self.collection, java:fromString(updateDoc), (), isMultiple, upsert);
+        }
+        string filterStr = filter.toJsonString();
+        return update(self.collection, java:fromString(updateDoc), java:fromString(filterStr), isMultiple, upsert);
+    }
+
+    # Delete document based on condition
```
Comment (praneesha) [inline mongodb/src/mongodb/collection.bal:94]:
```suggestion
    # Deletes a document based on a condition.
```

File: mongodb/src/mongodb/collection.bal [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+    #
+    # + filter - Filter for query
```
Comment (praneesha) [inline mongodb/src/mongodb/collection.bal:96]:
```suggestion
    # + filter - Filter for the query
```

File: mongodb/src/mongodb/collection.bal [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+    # + isMultiple - Delete multiple documents if condition match
```
Comment (praneesha) [inline mongodb/src/mongodb/collection.bal:97]:
```suggestion
    # + isMultiple - Delete multiple documents if the condition is matched
```

File: mongodb/src/mongodb/collection.bal [hunk 16]
```diff
@@ -0,0 +1,134 @@
...
+    # + return - number of deleted documents, else an error
```
Comment (praneesha) [inline mongodb/src/mongodb/collection.bal:98]:
```suggestion
    # + return - number of deleted documents or else returns an error
```

File: mongodb/src/mongodb/database.bal [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+import ballerina/java;
+
+public type Database client object {
+    handle database;
+
+    public function __init(handle database) {
+        self.database = database;
+    }
+
+    # List collection names in the MongoDB database.
```
Comment (praneesha) [inline mongodb/src/mongodb/database.bal:26]:
```suggestion
    # Lists the collection names in the MongoDB database.
```

File: mongodb/src/mongodb/database.bal [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+    #
+    # + return - A array of collection names on success else returns an error
```
Comment (praneesha) [inline mongodb/src/mongodb/database.bal:28]:
```suggestion
    # + return - An array of collection names on success or else returns an error
```

File: mongodb/src/mongodb/database.bal [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+    public remote function getCollectionNames() returns string[]|DatabaseError {
+        return getCollectionNames(self.database);
+    }
+
+    #
+    # + name - Name of the collection
+    # + return - A collection object on success, else returns an error
```
Comment (praneesha) [inline mongodb/src/mongodb/database.bal:35]:
```suggestion
    # + return - A collection object on success or else returns an error
```

File: mongodb/src/mongodb/errors.bal [hunk 20]
```diff
@@ -14,22 +14,37 @@
 // specific language governing permissions and limitations
 // under the License.
 
-# Holds the details of an MongoDB error
+# Holds the properties of an Application error
```
Comment (praneesha) [inline mongodb/src/mongodb/errors.bal:17]:
```suggestion
# Holds the properties of an application error.
```

File: mongodb/src/mongodb/errors.bal [hunk 21]
```diff
@@ -14,22 +14,37 @@
...
 # + cause - Cause of the error
-# + detail - Error detail
-public type ErrorDetail record {
+public type ApplicationErrorDetail record {
     string message;
     error cause?;
-    string detail?;
+};
+
+# Holds the properties of a Database error
```
Comment (praneesha) [inline mongodb/src/mongodb/errors.bal:26]:
```suggestion
# Holds the properties of a database error.
```

File: mongodb/src/mongodb/errors.bal [hunk 22]
```diff
@@ -14,22 +14,37 @@
...
+#
+# + message - Specific error message for the error
+# + mongoDBExceptionType - Type of the mongodb exception returned
```
Comment (praneesha) [inline mongodb/src/mongodb/errors.bal:29]:
```suggestion
# + mongoDBExceptionType - Type of the returned MongoDB exception 
```

File: mongodb/src/mongodb/errors.bal [hunk 23]
```diff
@@ -14,22 +14,37 @@
...
+# + cause - Cause of the error
+public type DatabaseErrorDetail record {
+    string message;
+    string mongoDBExceptionType;
+    error cause?;
 };
 
 // Error reasons
```
Comment (praneesha) [inline mongodb/src/mongodb/errors.bal:37]:
```suggestion
// Error reasons.
```

File: mongodb/src/mongodb/tests/Ballerina.md [hunk 24]
```diff
@@ -0,0 +1,21 @@
+# Compatibility
+
+| Ballerina Language Version  | MongoDB Version |
+| ----------------------------| -------------------------------|
+|  1.2.0                      |   4.2.0
+
+## Prerequisites
+
+1. MongoDB 4.2.0 running instance
```
Comment (praneesha) [inline mongodb/src/mongodb/tests/Ballerina.md:9]:
```suggestion
1. A MongoDB 4.2.0 running instance
```

File: mongodb/src/mongodb/tests/Ballerina.md [hunk 25]
```diff
@@ -0,0 +1,21 @@
...
+
+2. Create 'moviecollection' database in the MongoDB node.
```
Comment (praneesha) [inline mongodb/src/mongodb/tests/Ballerina.md:11]:
```suggestion
2. The 'moviecollection' database created in the MongoDB node.
```

File: mongodb/src/mongodb/tests/Ballerina.md [hunk 26]
```diff
@@ -0,0 +1,21 @@
...
+
+## Running the tests
+
+1. Configure `mongoConfig` in main_test.bal to add credentials and server address of the running MongoDB node
```
Comment (praneesha) [inline mongodb/src/mongodb/tests/Ballerina.md:15]:
```suggestion
1. Configure the `mongoConfig` in the `main_test.bal` file to add the credentials and server address of the running MongoDB node.
```

File: mongodb/src/mongodb/tests/Ballerina.md [hunk 27]
```diff
@@ -0,0 +1,21 @@
...
+
+2. Run the following command
```
Comment (praneesha) [inline mongodb/src/mongodb/tests/Ballerina.md:17]:
```suggestion
2. Execute the following command.
```

File: mongo-utils/src/main/java/org/wso2/mongo/MongoDBDataSourceUtil.java (hunk 1 above)
Comment (niveathika) [reply mongo-utils/src/main/java/org/wso2/mongo/MongoDBDataSourceUtil.java]: Fixed in edfb343

File: mongodb/src/mongodb/client.bal (hunk 2 above)
Comment (niveathika) [reply mongodb/src/mongodb/client.bal]: Fixed in 40c0f72

File: README.md [hunk 28]
```diff
@@ -1,102 +1,85 @@
 [![Build Status](https://travis-ci.org/ballerina-platform/module-mongodb.svg?branch=master)](https://travis-ci.org/ballerina-platform/module-mongodb)
 
-Ballerina MongoDB Client is used to connect Ballerina with MongoDB data source.<br/> <br/>
- With the Ballerina MongoDB client following operations are supported.
+Ballerina MongoDB Client is used to connect Ballerina with a MongoDB data source.
```
Comment (praneesha) [inline README.md:3]:
```suggestion
The Ballerina MongoDB Client is used to connect Ballerina with a MongoDB data source.
```

File: README.md [hunk 29]
```diff
@@ -1,102 +1,85 @@
...
+ The following operations are supported with the Ballerina MongoDB client.
```
Comment (praneesha) [inline README.md:4]:
```suggestion
 The following operations are supported by the Ballerina MongoDB client.
```

File: README.md [hunk 30]
```diff
@@ -1,102 +1,85 @@
...
-3. findOne - To select the first document match with the query.<br/>
-4. update - To update documents that matches to the given filter. <br/>
-5. delete - To delete documents that matches to the given filter.
-Steps to Configure <br/>
+1. getDatabaseNames - Get the database names in a given MongoDB node
+2. getCollectionNames - Get the collection names in a given MongoDB collection
+3. countDocuments - Count the documents in a given collection
+4. insert - To insert a document to a given collection
+5. find - To select a document from a given collection according to a given query
+6. update - To update documents, which matches the given filter
```
Comment (praneesha) [inline README.md:11]:
```suggestion
6. update - To update the documents, which matche the given filter
```

File: README.md [hunk 31]
```diff
@@ -1,102 +1,85 @@
...
+7. delete - To delete documents, which matches the given filter
```
Comment (praneesha) [inline README.md:12]:
```suggestion
7. delete - To delete the documents, which match the given filter
```

File: README.md [hunk 32]
```diff
@@ -1,102 +1,85 @@
...
 
 ## Samples
 
 ### Performing CRUD operations with MongoDB client
```
Comment (praneesha) [inline README.md:16]:
```suggestion
### Performing CRUD operations with a MongoDB client
```

File: README.md [hunk 33]
```diff
@@ -1,102 +1,85 @@
...
 
-Following is a simple Ballerina program that can be used to perform CRUD operations.
+The following is a simple Ballerina program that can be used to perform CRUD operations.
```
Comment (praneesha) [inline README.md:18]:
```suggestion
The following is a simple Ballerina program, which can be used to perform CRUD operations.
```

File: mongodb/src/mongodb/Module.md [hunk 34]
```diff
@@ -2,96 +2,92 @@ Connects to Mongo DB from Ballerina.
...
+| Ballerina Language          | 1.2.X                       |
 | Mongo DB                    | V4.2.0                      |
 
+## MongoDB Clients
+
+There are 3 clients provided by Ballerina to interact with MongoDB.
+
+1. **mongodb:Client** - This connects to the running MongoDB node and lists the database names as well as gets a client for a specific database.
+
+2. **mongodb:Database** - This connects to a specific MongoDB database and lists the collection names as well as get a client for a specific collection.
```
Comment (praneesha) [inline mongodb/src/mongodb/Module.md:20]:
```suggestion
2. **mongodb:Database** - This connects to a specific MongoDB database and lists the collection names as well as gets a client for a specific collection.
```

File: mongodb/src/mongodb/Module.md [hunk 35]
```diff
@@ -2,96 +2,92 @@ Connects to Mongo DB from Ballerina.
...
+
+3. **mongodb:Collection** - This connects to a specific collection and performs various operations such as `count`, `listIndexes`, `insert`, `find`, `update`, and `delete`.
+
 ## Sample
 
 First, import the `ballerina/mongodb` module into the ballerina project.
```
Comment (praneesha) [inline mongodb/src/mongodb/Module.md:26]:
```suggestion
First, import the `ballerina/mongodb` module into the Ballerina project.
```

File: mongodb/src/mongodb/client.bal [hunk 36]
```diff
@@ -0,0 +1,109 @@
...
+
+    handle datasource;
+
+    public function __init(ClientConfig config) returns ApplicationError? {
+        self.datasource = check initClient(config);
+    }
+
+    # Lists the database names in the MongoDB server.
+    #
+    # + return - An array of database names on success or else returns an error
```
Comment (praneesha) [inline mongodb/src/mongodb/client.bal:30]:
```suggestion
    # + return - An array of database names on success or else returns an error
```
Do we need to specify the error?

File: mongodb/src/mongodb/client.bal [hunk 37]
```diff
@@ -0,0 +1,109 @@
...
+    public remote function getDatabasesNames() returns string[]|DatabaseError {
+        return getDatabasesNames(self.datasource);
+    }
+
+    # Returns the `Database` client.
+    # 
+    # + name - Name of the database
+    # + return - A database client object on success or else returns an error
```
Comment (praneesha) [inline mongodb/src/mongodb/client.bal:38]:
```suggestion
    # + return - A database client object on success or else returns an error
```
Do we need to specify the error?

File: mongodb/src/mongodb/client.bal [hunk 38]
```diff
@@ -0,0 +1,109 @@
...
+function getDatabase(handle datasource, string databaseName) returns handle|Error = @java:Method {
+    class: "org.wso2.mongo.MongoDBDataSourceUtil"
+} external;
+
+function close(handle datasource) = @java:Method {
+    class: "org.wso2.mongo.MongoDBDataSourceUtil"
+} external;
+
+
+# The Client configuration for MongoDB.
```
Comment (praneesha) [inline mongodb/src/mongodb/client.bal:71]:
```suggestion
# The Client configurations for MongoDB.
```

File: mongodb/src/mongodb/collection.bal [hunk 39]
```diff
@@ -0,0 +1,134 @@
...
+        if (filter is ()) {
+            return countDocuments(self.collection, ());
+        }
+        string filterString = filter.toJsonString();
+        return countDocuments(self.collection, java:fromString(filterString));
+    }
+
+    # Lists the indices associated with the collection.
+    #
+    # + return - a JSON object with indices on success or else returns an error
```
Comment (praneesha) [inline mongodb/src/mongodb/collection.bal:40]:
```suggestion
    # + return - a JSON object with indices on success or else returns an error
```
Do we need to specify the error?

File: mongodb/src/mongodb/collection.bal [hunk 40]
```diff
@@ -0,0 +1,134 @@
...
+        string documentStr = document.toJsonString();
+        return insert(self.collection, java:fromString(documentStr));
+    }
+
+    # The queries collection for documents, which sorts and limits the returned results.
+    #
+    # + filter - Filter for the query
+    # + sort - Sort options for the query
+    # + limit - Limit options for the query results. No limit is applied for -1
+    # + return - JSON array of the documents in the collection or else returns an error
```
Comment (praneesha) [inline mongodb/src/mongodb/collection.bal:60]:
```suggestion
    # + return - JSON array of the documents in the collection or else returns an error
```
Do we need to specify the error?

File: mongodb/src/mongodb/collection.bal [hunk 41]
```diff
@@ -0,0 +1,134 @@
...
+        return find(self.collection, java:fromString(filterStr), java:fromString(sortString), limit);
+    }
+
+    # Updates a document based on a condition.
+    #
+    # + set - Document for the update condition
+    # + filter - Filter for the query
+    # + isMultiple - Whether to update multiple documents
+    # + upsert - Whether to insert if update cannot be achieved
+    # + return - JSON array of the documents in the collection or else returns an error
```
Comment (praneesha) [inline mongodb/src/mongodb/collection.bal:83]:
```suggestion
    # + return - JSON array of the documents in the collection or else returns an error
```
Do we need to specify the error?

File: mongodb/src/mongodb/collection.bal [hunk 42]
```diff
@@ -0,0 +1,134 @@
...
+        }
+        string filterStr = filter.toJsonString();
+        return update(self.collection, java:fromString(updateDoc), java:fromString(filterStr), isMultiple, upsert);
+    }
+
+    # Deletes a document based on a condition.
+    #
+    # + filter - Filter for the query
+    # + isMultiple - Delete multiple documents if the condition is matched
+    # + return - number of deleted documents or else returns an error
```
Comment (praneesha) [inline mongodb/src/mongodb/collection.bal:98]:
```suggestion
    # + return - The number of deleted documents or else returns an error
```
Do we need to specify the error?

File: mongodb/src/mongodb/database.bal [hunk 43]
```diff
@@ -0,0 +1,52 @@
...
+public type Database client object {
+    handle database;
+
+    public function __init(handle database) {
+        self.database = database;
+    }
+
+    # Lists the collection names in the MongoDB database.
+    #
+    # + return - An array of collection names on success or else returns an error
```
Comment (praneesha) [inline mongodb/src/mongodb/database.bal:28]:
```suggestion
    # + return - An array of collection names on success or else returns an error
```
Do we need to specify the error?

File: mongodb/src/mongodb/database.bal [hunk 44]
```diff
@@ -0,0 +1,52 @@
...
+    public remote function getCollectionNames() returns string[]|DatabaseError {
+        return getCollectionNames(self.database);
+    }
+
+    # Returns the collection object.
+    # 
+    # + name - Name of the collection
+    # + return - A collection object on success or else returns an error
```
Comment (praneesha) [inline mongodb/src/mongodb/database.bal:36]:
```suggestion
    # + return - A collection object on success or else returns an error
```
Do we need to specify the error?

File: mongodb/src/mongodb/errors.bal [hunk 45]
```diff
@@ -14,22 +14,37 @@
...
+
 public const APPLICATION_ERROR_REASON = "{ballerina/mongodb}ApplicationError";
 
-public type DatabaseError error<DATABASE_ERROR_REASON, ErrorDetail>;
-public type ApplicationError error<APPLICATION_ERROR_REASON, ErrorDetail>;
+# Represents an error originating from application-level causes.
+public type ApplicationError error<APPLICATION_ERROR_REASON, ApplicationErrorDetail>;
 
-public type ConnectorError DatabaseError|ApplicationError;
+# Represents a database or application level error returned from MongoDB client remote functions.
```
Comment (praneesha) [inline mongodb/src/mongodb/errors.bal:49]:
```suggestion
# Represents a database or application level error returned from the MongoDB client remote functions.
```

File: mongodb/src/mongodb/tests/Ballerina.md [hunk 46]
```diff
@@ -0,0 +1,22 @@
...
+
+1. A MongoDB 4.2.0 running instance
+
+2. The 'moviecollection' database created in the MongoDB node.
+
+## Running the tests
+
+1. Configure the `mongoConfig` in the `main_test.bal` file to add the credentials and server address of the running MongoDB node.
+
+2. Execute the following commands inside the repo root folder.
```
Comment (praneesha) [inline mongodb/src/mongodb/tests/Ballerina.md:17]:
```suggestion
2. Execute the following commands inside the root folder of the GitHub repo.
```

File: mongodb/src/mongodb/client.bal (hunk 36 above)
Comment (niveathika) [reply mongodb/src/mongodb/client.bal]: Yes as there can be error due t network when trying t connect t the DB Improved doc comment in c8145ff

File: mongodb/src/mongodb/client.bal (hunk 37 above)
Comment (niveathika) [reply mongodb/src/mongodb/client.bal]: Same Fixed in c8145ff

File: mongodb/src/mongodb/collection.bal (hunk 39 above)
Comment (niveathika) [reply mongodb/src/mongodb/collection.bal]: Same Fixed in c8145ff

File: mongodb/src/mongodb/collection.bal (hunk 40 above)
Comment (niveathika) [reply mongodb/src/mongodb/collection.bal]: Same Fixed in c8145ff

File: mongodb/src/mongodb/collection.bal (hunk 41 above)
Comment (niveathika) [reply mongodb/src/mongodb/collection.bal]: Same Fixed in c8145ff

File: mongodb/src/mongodb/database.bal (hunk 43 above)
Comment (niveathika) [reply mongodb/src/mongodb/database.bal]: Same Fixed in c8145ff

File: mongodb/src/mongodb/database.bal (hunk 44 above)
Comment (niveathika) [reply mongodb/src/mongodb/database.bal]: Same Fixed in c8145ff

**Pull request merged**